spot-vaults/contracts/BillBroker.sol

```diff
@@ -0,0 +1,753 @@
+        //   fee
+        //    ^
+        //    |
+        // fh |         |           /
+        //    |         |          /
+        //    |         |         /
+        //    |         |        /
+        //    |         |       /
+        //    |         |      /
+        // fl |     __________/
+        //    |         |
+        //    |         |
+        //    |         |
+        //    +---------------------------> ar
+        //              1.0  arSU   arHU
+        //
+        Range memory swapFeePercs = fees.usdToPerpSwapFeePercs;
+        return
+            _computeFeePerc(
+                LinearFn({
+                    x1: 0,
+                    y1: swapFeePercs.lower,
+                    x2: ONE,
+                    y2: swapFeePercs.lower
+                }),
+                LinearFn({
+                    x1: arSoftBound.upper,
+                    y1: swapFeePercs.lower,
+                    x2: arHardBound.upper,
+                    y2: swapFeePercs.upper
+                }),
+                arPre,
+                arPost,
+                arSoftBound.upper
+            );
+    }
+
+    /// @param s The system reserve state.
+    /// @return The computed asset ratio of the system.
+    function assetRatio(ReserveState memory s) public view returns (uint256) {
+        return
+            s.usdReserve.mulDiv(s.usdPrice, usdUnitAmt).mulDiv(
+                ONE,
+                s.perpReserve.mulDiv(s.perpPrice, perpUnitAmt)
+            );
+    }
+
+    /// @return The balance of usd tokens in the reserve.
+    function usdReserve() public view returns (uint256) {
+        return usd.balanceOf(address(this));
+    }
+
+    /// @return The balance of perp tokens in the reserve.
+    function perpReserve() public view returns (uint256) {
+        return perp.balanceOf(address(this));
+    }
+
+    //-----------------------------------------------------------------------------
+    // Private methods
+
+    /// @dev The function assumes the fee curve is defined a pair-wise linear function which merge at the cutoff point.
+    ///      The swap fee is computed as area under the fee curve between {arL,arU}.
+    function _computeFeePerc(
+        LinearFn memory fn1,
+        LinearFn memory fn2,
+        uint256 arL,
+        uint256 arU,
+        uint256 cutoff
+    ) private pure returns (uint256) {
+        if (arU <= cutoff) {
+            return _auc(fn1, arL, arU);
+        } else if (arL > cutoff) {
+            return _auc(fn2, arL, arU);
+        } else {
+            return (_auc(fn1, arL, cutoff).mulDiv(cutoff - arL, arU - arL) +
+                _auc(fn2, cutoff, arU).mulDiv(arU - cutoff, arU - arL));
+        }
+    }
+
+    /// @dev Given a linear function defined by points (x1,y1) (x2,y2),
+    ///      we compute the are under the curve between (xL, xU) assuming xL <= xU.
+    function _auc(
+        LinearFn memory fn,
+        uint256 xL,
+        uint256 xU
+    ) private pure returns (uint256) {
```

Comment:
Since this is an unsigned linear function, could it be simpler to do it like a trapezoid?

(xU - xL) * (y2+y1) / 2

Reply:
Right, that's eventually what it will cancel out into